kbase-extension/static/kbase/js/widgets/narrative_core/upload/importErrors.js

```diff
@@ -0,0 +1,246 @@
+            this.serverErrors = [];
+            this.unexpectedErrors = [];
+            this.noFileError = false;
+            this._processErrors();
+        }
+
+        toString() {
+            return `${this.name} - ${this.text}: ${JSON.stringify(this.errors)}`;
+        }
+
+        /**
+         * This takes the list of errors, sorts, and groups them by file type where possible.
+         * The goal is to convert from a list of errors of various formats to something more
+         * easily printable with some minor formatting.
+         *
+         * i.e. from:
+         * [{
+         *   type: 'cannot_find_file',
+         *   file: 'some_file.csv',
+         * }]
+         * to:
+         * {
+         *   some_file.csv: ['not found']
+         * }
+         * or, more complicated:
+         * [{
+         *   type: 'incorrect_column_count',
+         *   file: 'file1.xls',
+         *   tab: 'tab_1',
+         *   message: 'wrong column count, missing <some column>'
+         * }, {
+         *   type: 'multiple_specifications_for_data_type',
+         *   file_1: 'file1.xls',
+         *   tab_1: 'tab_1',
+         *   file_2: 'file2.csv',
+         *   tab_2: null,
+         *   message: 'duplicate specification found'
+         * }, {
+         *   type: 'unexpected_error',
+         *   message: 'an unexpected error occurred'
+         * }]
+         * to:
+         * {
+         *   'file1.xls tab "tab_1"': [
+         *     'wrong column count, missing <some column>',
+         *     'duplicate specification found in file2.csv'
+         *   ],
+         *   'file2.csv': [
+         *     'duplicate specification found in file1.xls tab "tab_1"'
+         *   ],
+         * }
+         */
+        _processErrors() {
+            // a little helper to handle file error formatting
+            const addFileError = (error) => {
+                let fileKey = error.file;
+                if (error.tab) {
+                    fileKey += ` tab "${error.tab}"`;
+                }
+                if (!this.fileErrors[fileKey]) {
+                    this.fileErrors[fileKey] = [];
+                }
+                this.fileErrors[fileKey].push(error.message);
+            };
+
+            this.errors.forEach((error) => {
+                switch (error.type) {
+                    case BULK_SPEC_ERRORS.CANNOT_PARSE:
+                    case BULK_SPEC_ERRORS.INCORRECT_COLUMN_COUNT:
+                        addFileError(error);
+                        break;
+                    case BULK_SPEC_ERRORS.NOT_FOUND:
+                        addFileError(Object.assign({ message: 'File not found' }, error));
+                        break;
+                    case BULK_SPEC_ERRORS.MULTIPLE_SPECS:
+                        addFileError({
+                            file: error.file_1,
+                            tab: error.tab_1,
+                            message: error.message,
+                        });
```

Comment:
This is out of scope for this particular PR, but how does the staging service (presumably) handle cases where a spec appears in more than two files?

Reply:
That's a good question. It makes multiple errors, not necessarily involving all combinations, though.
Example: I tried to force an upload of 4 files that are all assembly specs, one's an excel with an assembly tab. Here's the result:

```
{
    "errors": [
        {
            "type": "multiple_specifications_for_data_type",
            "message": "Data type assembly appears in two importer specification sources",
            "file_1": "wjriehl/excel_uploads.xlsx",
            "tab_1": "assembly",
            "file_2": "wjriehl/assembly_test.csv",
            "tab_2": null
        },
        {
            "type": "multiple_specifications_for_data_type",
            "message": "Data type assembly appears in two importer specification sources",
            "file_1": "wjriehl/excel_uploads.xlsx",
            "tab_1": "assembly",
            "file_2": "wjriehl/assembly_test2.csv",
            "tab_2": null
        },
        {
            "type": "multiple_specifications_for_data_type",
            "message": "Data type assembly appears in two importer specification sources",
            "file_1": "wjriehl/excel_uploads.xlsx",
            "tab_1": "assembly",
            "file_2": "wjriehl/assembly_test_3.csv",
            "tab_2": null
        }
    ]
}
```

`excel_uploads.xlsx` was the first in the list, so I think they're all just relative to that.

Reply:
And I agree that case needs a little more management, but maybe not yet? It might be worth making a ticket to come back to once other core error cases are handled.

Reply:
Thanks for the investigatory work. It looks like the current coding would display those errors in a reasonable manner, so it's fine for now. 

This might be better handled by having the staging service collect the file/tab error information and supply a list of all related duplicates, rather than repeating the info about the first file with each subsequent duplication.